cpp/ql/src/experimental/Security/CWE/CWE-193/ConstantSizeArrayOffByOne.ql

```diff
@@ -56,9 +57,9 @@ predicate isFieldAddressSource(Field f, DataFlow::Node source) {
  */
 predicate isInvalidPointerDerefSink(DataFlow::Node sink, Instruction i, string operation) {
   exists(AddressOperand addr, int delta |
-    bounded(addr.getDef(), sink.asInstruction(), delta) and
+    bounded(pragma[only_bind_into](addr).getDef(), sink.asInstruction(), delta) and
```

Comment:
I'm surprised we don't need to inline this predicate like it seems we need to in the `cpp/invalid-pointer-deref` query 🤔. Does the compiler inline the predicate in this query automatically, but not in the other query?

Reply:
It didn't inline it. The tuple counts are still quite substantial here, so inlining might improve it further:
```
[2023-04-06 11:25:49] Evaluated non-recursive predicate ConstantSizeArrayOffByOne#f6a3dbf4::isInvalidPointerDerefSink#3#fff@c640311h in 20054ms (size: 145734470).
Evaluated relational algebra for predicate ConstantSizeArrayOffByOne#f6a3dbf4::isInvalidPointerDerefSink#3#fff@c640311h with tuple counts:
          5205652   ~0%    {3} r1 = JOIN Instruction#577b6a83::Instruction::getAnOperand#0#dispred#ff WITH Instruction#577b6a83::LoadInstruction#f ON FIRST 1 OUTPUT Lhs.1, Lhs.0, "read"
                       
          1938182   ~0%    {3} r2 = JOIN Instruction#577b6a83::Instruction::getAnOperand#0#dispred#ff WITH Instruction#577b6a83::StoreInstruction#f ON FIRST 1 OUTPUT Lhs.1, Lhs.0, "write"
                       
          7143834   ~0%    {3} r3 = r1 UNION r2
          4541008   ~0%    {4} r4 = JOIN r3 WITH Operand#27b358ab::Operand::getDef#0#dispred#ff ON FIRST 1 OUTPUT Lhs.0, Lhs.1, Lhs.2, Rhs.1
          3571886   ~0%    {3} r5 = JOIN r4 WITH Operand#27b358ab::AddressOperand#ffff ON FIRST 1 OUTPUT Lhs.3, Lhs.1, Lhs.2
          3571754   ~4%    {4} r6 = JOIN r5 WITH EquivalenceRelation#Instruction#577b6a83::Instruction#SemanticExprSpecific#ad5ae873::SemanticExprConfig::idOrSafeConversion#::getEquivalenceClass#ff ON FIRST 1 OUTPUT Lhs.0, Lhs.1, Lhs.2, Rhs.1
          3571754   ~2%    {4} r7 = JOIN r6 WITH SSAConstruction#2b11997e::Cached::getInstructionEnclosingIRFunction#1#ff ON FIRST 1 OUTPUT Lhs.3, Lhs.1, Lhs.2, Rhs.1
          3604932   ~0%    {5} r8 = JOIN r7 WITH project#RangeAnalysisImpl#edd69a76::semBounded#5#fffbf ON FIRST 1 OUTPUT Lhs.1, Lhs.2, Lhs.3, Rhs.1, f2i(Rhs.2)
          3604644   ~0%    {5} r9 = SELECT r8 ON In.4 >= 0
          3604644   ~0%    {4} r10 = SCAN r9 OUTPUT In.3, In.2, In.0, In.1
        145736417   ~0%    {3} r11 = JOIN r10 WITH ConstantSizeArrayOffByOne#f6a3dbf4::getABoundIn#2#fff ON FIRST 2 OUTPUT Rhs.2, Lhs.2, Lhs.3
        145734470   ~6%    {3} r12 = JOIN r11 WITH DataFlowUtil#47741e1f::Node::asInstruction#0#dispred#ff_10#join_rhs ON FIRST 1 OUTPUT Rhs.1, Lhs.1, Lhs.2
                           return r12

```

Reply:
... or rather _up_ the chain of calls, I guess 😂

Reply:
With some inlining I get:
```
[2023-04-06 11:52:59] Evaluated non-recursive predicate DataFlowImpl#9021cc4c::DefaultState#ConstantSizeArrayOffByOne#f6a3dbf4::PointerArithmeticToDerefConfig#::isSink#2#cpe#1#f@9f6f426r in 3369ms (size: 3895585).
Evaluated relational algebra for predicate DataFlowImpl#9021cc4c::DefaultState#ConstantSizeArrayOffByOne#f6a3dbf4::PointerArithmeticToDerefConfig#::isSink#2#cpe#1#f@9f6f426r with tuple counts:
         5205652    ~5%    {1} r1 = JOIN Instruction#577b6a83::Instruction::getAnOperand#0#dispred#ff WITH Instruction#577b6a83::LoadInstruction#f ON FIRST 1 OUTPUT Lhs.1
                       
         1938182    ~0%    {1} r2 = JOIN Instruction#577b6a83::Instruction::getAnOperand#0#dispred#ff WITH Instruction#577b6a83::StoreInstruction#f ON FIRST 1 OUTPUT Lhs.1
                       
         7143834    ~3%    {1} r3 = r1 UNION r2
         3571886    ~1%    {1} r4 = JOIN r3 WITH Operand#27b358ab::AddressOperand#ffff ON FIRST 1 OUTPUT Lhs.0
         3571886    ~9%    {1} r5 = JOIN r4 WITH Operand#27b358ab::Operand::getDef#0#dispred#ff ON FIRST 1 OUTPUT Rhs.1
         3571754    ~4%    {2} r6 = JOIN r5 WITH EquivalenceRelation#Instruction#577b6a83::Instruction#SemanticExprSpecific#ad5ae873::SemanticExprConfig::idOrSafeConversion#::getEquivalenceClass#ff ON FIRST 1 OUTPUT Lhs.0, Rhs.1
         3571754    ~3%    {2} r7 = JOIN r6 WITH SSAConstruction#2b11997e::Cached::getInstructionEnclosingIRFunction#1#ff ON FIRST 1 OUTPUT Lhs.1, Rhs.1
         3604932  ~136%    {3} r8 = JOIN r7 WITH project#RangeAnalysisImpl#edd69a76::semBounded#5#fffbf ON FIRST 1 OUTPUT Lhs.1, Rhs.1, f2i(Rhs.2)
         3604644  ~136%    {3} r9 = SELECT r8 ON In.2 >= 0
         3604644  ~147%    {2} r10 = SCAN r9 OUTPUT In.1, In.0
        13463637  ~255%    {1} r11 = JOIN r10 WITH ConstantSizeArrayOffByOne#f6a3dbf4::getABoundIn#2#fff ON FIRST 2 OUTPUT Rhs.2
        13457649  ~252%    {1} r12 = JOIN r11 WITH DataFlowUtil#47741e1f::Node::asInstruction#0#dispred#ff_10#join_rhs ON FIRST 1 OUTPUT Rhs.1
                           return r12
```
and
```
[2023-04-06 11:53:10] Evaluated non-recursive predicate #select#ffffffff@9f6a7bmo in 10409ms (size: 10).
Evaluated relational algebra for predicate #select#ffffffff@9f6a7bmo with tuple counts:
          7156311   ~0%    {2} r1 = JOIN Instruction#577b6a83::Instruction::getAnOperand#0#dispred#ff_10#join_rhs WITH Operand#27b358ab::AddressOperand#ffff ON FIRST 1 OUTPUT Lhs.0, Lhs.1
          7156311   ~0%    {2} r2 = JOIN r1 WITH Operand#27b358ab::Operand::getDef#0#dispred#ff ON FIRST 1 OUTPUT Rhs.1, Lhs.1
          7156179   ~0%    {3} r3 = JOIN r2 WITH EquivalenceRelation#Instruction#577b6a83::Instruction#SemanticExprSpecific#ad5ae873::SemanticExprConfig::idOrSafeConversion#::getEquivalenceClass#ff ON FIRST 1 OUTPUT Lhs.0, Lhs.1, Rhs.1
          7156179   ~3%    {3} r4 = JOIN r3 WITH SSAConstruction#2b11997e::Cached::getInstructionEnclosingIRFunction#1#ff ON FIRST 1 OUTPUT Lhs.2, Lhs.1, Rhs.1
          7190602   ~1%    {4} r5 = JOIN r4 WITH project#RangeAnalysisImpl#edd69a76::semBounded#5#fffbf ON FIRST 1 OUTPUT Lhs.1, Lhs.2, Rhs.1, f2i(Rhs.2)
          7190201   ~1%    {4} r6 = SELECT r5 ON In.3 >= 0
          7190201   ~0%    {3} r7 = SCAN r6 OUTPUT In.2, In.1, In.0
        272160850   ~0%    {2} r8 = JOIN r7 WITH ConstantSizeArrayOffByOne#f6a3dbf4::getABoundIn#2#fff ON FIRST 2 OUTPUT Rhs.2, Lhs.2
        272103824   ~0%    {2} r9 = JOIN r8 WITH DataFlowUtil#47741e1f::Node::asInstruction#0#dispred#ff_10#join_rhs ON FIRST 1 OUTPUT Rhs.1, Lhs.1
               10   ~0%    {2} r10 = JOIN r9 WITH DataFlowImpl#9021cc4c::Impl#DataFlow#6c4a6647::Global#ConstantSizeArrayOffByOne#f6a3dbf4::PointerArithmeticToDerefConfig#::C#::PathNode::getNode#0#dispred#ff_10#join_rhs ON FIRST 1 OUTPUT Rhs.1, Lhs.1
               10   ~0%    {3} r11 = JOIN r10 WITH DataFlowImpl#9021cc4c::Impl#DataFlow#6c4a6647::Global#ConstantSizeArrayOffByOne#f6a3dbf4::PointerArithmeticToDerefConfig#::C#::flowPath#2#ff_10#join_rhs ON FIRST 1 OUTPUT Rhs.1, Lhs.1, Lhs.0
               10   ~0%    {5} r12 = JOIN r11 WITH DataFlowImpl#9021cc4c::Impl#DataFlow#6c4a6647::Global#ConstantSizeArrayOffByOne#f6a3dbf4::PointerArithmeticToDerefConfig#::C#::PathNode::getNode#0#dispred#ff ON FIRST 1 OUTPUT Rhs.1, Lhs.1, Lhs.2, Lhs.0, Lhs.0
               10   ~0%    {5} r13 = JOIN r12 WITH DataFlowUtil#47741e1f::Node::asInstruction#0#dispred#ff ON FIRST 1 OUTPUT Rhs.1, Lhs.1, Lhs.2, Lhs.3, Lhs.4
               10   ~0%    {6} r14 = JOIN r13 WITH ConstantSizeArrayOffByOne#f6a3dbf4::isConstantSizeOverflowSource#3#fff_102#join_rhs ON FIRST 1 OUTPUT Rhs.1, Lhs.1, Lhs.2, Lhs.3, Lhs.4, Rhs.2
               10   ~0%    {7} r15 = JOIN r14 WITH Variable#7a968d4e::MemberVariable::getName#0#dispred#ff ON FIRST 1 OUTPUT Lhs.1, Lhs.2, Lhs.3, Lhs.4, Lhs.0, Rhs.1, ("This pointer arithmetic may have an off-by-" ++ toString((Lhs.5 + 1)) ++ " error allowing it to overrun $@ at this $@.")
                       
                8   ~0%    {8} r16 = JOIN r15 WITH Instruction#577b6a83::LoadInstruction#f ON FIRST 1 OUTPUT Lhs.2, Lhs.3, Lhs.1, Lhs.6, Lhs.4, Lhs.5, Lhs.0, "read"
                       
                2   ~0%    {8} r17 = JOIN r15 WITH Instruction#577b6a83::StoreInstruction#f ON FIRST 1 OUTPUT Lhs.2, Lhs.3, Lhs.1, Lhs.6, Lhs.4, Lhs.5, Lhs.0, "write"
                       
               10   ~0%    {8} r18 = r16 UNION r17
                           return r18
```

Reply:
Can you also inline `isSink`? That's what I did in https://github.com/MathiasVP/ql/commit/c5bb7fa59d83fe7dbcc7c2ad492756160f0b7eee.

Reply:
Doing that already.

Reply:
Hm, that's weird. It doesn't look like it actually inlined `PointerArithmeticToDerefConfig`'s `isSink`. I wonder if it _did_ inline it, but then materialized a cartesian-product-eliminated version of the predicate (i.e., the `#cpe`-suffixed predicate).

Reply:
Let me re-run to see if I looked at the right summary.

Reply:
Also for the final `select`: Could we change the order so that we started with `flowPath`?

Reply:
Tried that, which made things worse (I'll get you the tuple counts)

Reply:
😭

Reply:
> Let me re-run to see if I looked at the right summary.

Yeah, looked at the right summary.

> Tried that, which made things worse (I'll get you the tuple counts)

With `  PointerArithmeticToDerefFlow::flowPath(pragma[only_bind_into](source), pragma[only_bind_into](sink))`:
```
[2023-04-06 12:38:25] Evaluated non-recursive predicate #select#ffffffff@92a5fek5 in 19764ms (size: 10).
Evaluated relational algebra for predicate #select#ffffffff@92a5fek5 with tuple counts:
                6   ~0%    {2} r1 = SCAN DataFlowImpl#9021cc4c::Impl#DataFlow#6c4a6647::Global#ConstantSizeArrayOffByOne#f6a3dbf4::PointerArithmeticToDerefConfig#::C#::flowPath#2#ff OUTPUT In.1, In.0
                6   ~0%    {3} r2 = JOIN r1 WITH DataFlowImpl#9021cc4c::Impl#DataFlow#6c4a6647::Global#ConstantSizeArrayOffByOne#f6a3dbf4::PointerArithmeticToDerefConfig#::C#::PathNode::getNode#0#dispred#ff ON FIRST 1 OUTPUT Lhs.1, Lhs.0, Rhs.1
                6   ~0%    {5} r3 = JOIN r2 WITH DataFlowImpl#9021cc4c::Impl#DataFlow#6c4a6647::Global#ConstantSizeArrayOffByOne#f6a3dbf4::PointerArithmeticToDerefConfig#::C#::PathNode::getNode#0#dispred#ff ON FIRST 1 OUTPUT Rhs.1, Lhs.0, Lhs.1, Lhs.2, Lhs.0
                6   ~0%    {5} r4 = JOIN r3 WITH DataFlowUtil#47741e1f::Node::asInstruction#0#dispred#ff ON FIRST 1 OUTPUT Lhs.3, Lhs.1, Lhs.2, Lhs.4, Rhs.1
                6   ~0%    {5} r5 = JOIN r4 WITH DataFlowUtil#47741e1f::Node::asInstruction#0#dispred#ff ON FIRST 1 OUTPUT Lhs.4, Lhs.1, Lhs.2, Lhs.3, Rhs.1
                6   ~0%    {6} r6 = JOIN r5 WITH ConstantSizeArrayOffByOne#f6a3dbf4::isConstantSizeOverflowSource#3#fff_102#join_rhs ON FIRST 1 OUTPUT Rhs.1, Lhs.1, Lhs.2, Lhs.3, Lhs.4, Rhs.2
                6   ~0%    {7} r7 = JOIN r6 WITH Variable#7a968d4e::MemberVariable::getName#0#dispred#ff ON FIRST 1 OUTPUT Lhs.1, Lhs.2, Lhs.3, Lhs.4, Lhs.0, Rhs.1, ("This pointer arithmetic may have an off-by-" ++ toString((Lhs.5 + 1)) ++ " error allowing it to overrun $@ at this $@.")
        137026752   ~0%    {9} r8 = JOIN r7 WITH Instruction#577b6a83::Instruction::getAnOperand#0#dispred#ff CARTESIAN PRODUCT OUTPUT Rhs.1, Lhs.0, Lhs.1, Lhs.2, Lhs.3, Lhs.4, Lhs.5, Lhs.6, Rhs.0
         42937866   ~0%    {9} r9 = JOIN r8 WITH Operand#27b358ab::AddressOperand#ffff ON FIRST 1 OUTPUT Lhs.0, Lhs.1, Lhs.2, Lhs.3, Lhs.4, Lhs.5, Lhs.6, Lhs.7, Lhs.8
         42937866   ~0%    {9} r10 = JOIN r9 WITH Operand#27b358ab::Operand::getDef#0#dispred#ff ON FIRST 1 OUTPUT Rhs.1, Lhs.1, Lhs.2, Lhs.3, Lhs.4, Lhs.5, Lhs.6, Lhs.7, Lhs.8
         42937074   ~0%    {10} r11 = JOIN r10 WITH EquivalenceRelation#Instruction#577b6a83::Instruction#SemanticExprSpecific#ad5ae873::SemanticExprConfig::idOrSafeConversion#::getEquivalenceClass#ff ON FIRST 1 OUTPUT Lhs.0, Lhs.1, Lhs.2, Lhs.3, Lhs.4, Lhs.5, Lhs.6, Lhs.7, Lhs.8, Rhs.1
         42937074   ~0%    {10} r12 = JOIN r11 WITH SSAConstruction#2b11997e::Cached::getInstructionEnclosingIRFunction#1#ff ON FIRST 1 OUTPUT Lhs.9, Lhs.1, Lhs.2, Lhs.3, Lhs.4, Lhs.5, Lhs.6, Lhs.7, Lhs.8, Rhs.1
         43143612   ~0%    {11} r13 = JOIN r12 WITH project#RangeAnalysisImpl#edd69a76::semBounded#5#fffbf ON FIRST 1 OUTPUT Lhs.1, Lhs.2, Lhs.3, Lhs.4, Lhs.5, Lhs.6, Lhs.7, Lhs.8, Lhs.9, Rhs.1, f2i(Rhs.2)
         43141206   ~0%    {11} r14 = SELECT r13 ON In.10 >= 0
         43141206   ~0%    {10} r15 = SCAN r14 OUTPUT In.9, In.8, In.3, In.0, In.1, In.2, In.4, In.5, In.6, In.7
               10   ~0%    {7} r16 = JOIN r15 WITH ConstantSizeArrayOffByOne#f6a3dbf4::getABoundIn#2#fff ON FIRST 3 OUTPUT Lhs.9, Lhs.3, Lhs.4, Lhs.5, Lhs.6, Lhs.7, Lhs.8
                       
                8   ~0%    {8} r17 = JOIN r16 WITH Instruction#577b6a83::LoadInstruction#f ON FIRST 1 OUTPUT Lhs.1, Lhs.3, Lhs.2, Lhs.6, Lhs.4, Lhs.5, Lhs.0, "read"
                       
                2   ~0%    {8} r18 = JOIN r16 WITH Instruction#577b6a83::StoreInstruction#f ON FIRST 1 OUTPUT Lhs.1, Lhs.3, Lhs.2, Lhs.6, Lhs.4, Lhs.5, Lhs.0, "write"
                       
               10   ~0%    {8} r19 = r17 UNION r18
                           return r19
```
Tuple counts are slightly better, but it gets a bout twice as slow.

Reply:
Yeah, no wonder it gets slow:
```
137026752   ~0%    {9} r8 = JOIN r7 WITH Instruction#577b6a83::Instruction::getAnOperand#0#dispred#ff CARTESIAN PRODUCT OUTPUT Rhs.1, Lhs.0, Lhs.1, Lhs.2, Lhs.3, Lhs.4, Lhs.5, Lhs.6, Rhs.0
```
I think we're getting a cartesian product here because:
1. Due to the `bindingset[i]` annotation on `bounded`, the `pragma[only_bind_into](addr).getDef()` argument must be bounded before joining with `bounded`.
2. Due to the `only_bind_into` annotation on the argument, it must be bound by the joining with the `bounded` predicate.

The result: No join is possible, so we resort to a cartesian product.

I imagine you can also see this by running `codeql query compile .... --dump-dil` and observe that the `Instruction::getAnOperand` conjunct is mentioning several not-yet-bound variables without relating them to any other already-bound variables.

I wonder if we can just remove that `pragma[only_bind_into]` pragma?

Reply:
Oh, looks like you _just_ did in 8e7eec173be425a3725e1fb3368c06fa26074627 😂

Reply:
Yeah, I removed it, because it no longer had any effect after inlining. That didn't change the above numbers though.

Reply:
And can we still not order `flowPath` first in the select? 😕

Reply:
No, the numbers above were with the changes from https://github.com/github/codeql/commit/8e7eec173be425a3725e1fb3368c06fa26074627